PeakNetwork/Core/DecodeInputOperation.swift

```diff
@@ -0,0 +1,242 @@
+//
+//  DecodingOperation.swift
+//  PeakNetwork
+//
+//  Created by Sam Oakley on 12/02/2019.
+//  Copyright © 2019 3Squared. All rights reserved.
+//
+
+#if os(iOS) || os(tvOS)
+import UIKit
+#else
+import AppKit
+#endif
+import PeakOperation
+import PeakResult
+
+
+/// An operation which takes an Input and decodes it to an Output.
+/// A more specialised version of `MapOperation`, easier to use for this specific task.
+open class DecodeInputOperation<I, O>: ConcurrentOperation, ConsumesResult, ProducesResult {
+    
+    public var input: Result<I> = Result { throw ResultError.noResult }
+    public var output: Result<O> = Result { throw ResultError.noResult }
+
+    /// Create a new `DecodeInputOperation`.
+    ///
+    /// - Parameter input: An optional input value.
+    public init(input: I? = nil) {
+        super.init()
+        if let input = input {
+            self.input = .success(input)
+        }
+    }
+    
+    open override func execute() {
+        output = input.fold({ response in
+            return self.decode(input: response)
+        }, { error in
+            return .failure(error)
+        })
+        
+        finish()
+    }
+
+    
+    /// Convert a `success` input into a new output Result.
+    ///
+    /// - Parameter input:
+    /// - Returns: A converted output Result.
+    open func decode(input: I) -> Result<O> {
+        fatalError("Subclasses must implement `decode(::)`.")
+    }
+}
+
+
+/// Decode a network response using a `JSONDecoder`.
+open class JSONDecodeOperation<D: Decodable>: DecodeInputOperation<NetworkResponse, D> {
```

Comment:
Perhaps this could instead be done as a SequenceOperation that uses a JSONDecodeResponseOperation and removes the Response part of the Result.

Reply:
Makes it more complicated than it needs to.